File: cpp/basics/variables.cpp
/**
 * C++ Basics - Variables và Data Types
 */
#include <iostream>
#include <string>

using namespace std;

int main() {
    // 1. Primitive Data Types
    
    // Integer types
    int integer = 42;
    short shortInt = 32767;
    long longInt = 2147483647L;
    long long longLongInt = 9223372036854775807LL;
    
    // Floating point types
    float floatNum = 3.14f;
    double doubleNum = 3.14159;
    long double longDouble = 3.141592653589793238L;
    
    // Character types
    char character = 'A';
    wchar_t wideChar = L'A';
    
    // Boolean type
    bool isTrue = true;
    bool isFalse = false;
    
    // 2. String
    string str = "Hello, C++!";
    
    // 3. Arrays
    int numbers[5] = {1, 2, 3, 4, 5};
    int numbers2[] = {1, 2, 3};
    
    // 4. Pointers
    int x = 10;
    int* ptr = &x;
    cout << "Value: " << x << endl;
    cout << "Address: " << ptr << endl;
    cout << "Value via pointer: " << *ptr << endl;
    
    // 5. References
    int y = 20;
    int& ref = y;
    ref = 30;
    cout << "y = " << y << endl;  // y is now 30
    
    // 6. Constants
    const int MAX_SIZE = 100;
    const double PI = 3.14159;
    
    // 7. Auto keyword (C++11)
    auto autoInt = 42;
    auto autoDouble = 3.14;
    auto autoString = "Hello";
    
    // 8. Output
    cout << "Integer: " << integer << endl;
    cout << "Double: " << doubleNum << endl;
    cout << "String: " << str << endl;
    cout << "Boolean: " << isTrue << endl;
    
    return 0;
}

File: java/basics/Variables.java
/**
 * Java Basics - Variables và Data Types
 */
public class Variables {
    public static void main(String[] args) {
        // 1. Primitive Data Types
        
        // Integer types
        byte byteVar = 127;           // 8-bit, -128 to 127
        short shortVar = 32767;       // 16-bit
        int intVar = 2147483647;      // 32-bit
        long longVar = 9223372036854775807L; // 64-bit
        
        // Floating point types
        float floatVar = 3.14f;       // 32-bit
        double doubleVar = 3.14159;   // 64-bit
        
        // Character type
        char charVar = 'A';
        
        // Boolean type
        boolean boolVar = true;
        
        // 2. Reference Data Types
        String str = "Hello, Java!";
        String str2 = new String("Hello");
        
        // 3. Arrays
        int[] numbers = {1, 2, 3, 4, 5};
        int[] numbers2 = new int[5];
        
        // 4. Type conversion
        int x = 10;
        double y = x;  // Implicit conversion
        
        double d = 3.14;
        int i = (int) d;  // Explicit cast
        
        // 5. Constants
        final int MAX_SIZE = 100;
        final String APP_NAME = "MyApp";
        
        // 6. Print variables
        System.out.println("Integer: " + intVar);
        System.out.println("Double: " + doubleVar);
        System.out.println("String: " + str);
        System.out.println("Boolean: " + boolVar);
    }
}

